rpm/0007-Add-RFC7636-aka-PKCE-support.patch

```diff
@@ -0,0 +1,133 @@
+SkipPKCE property that can be set to true.
+
+There is no support for "plain" method as the specification requires
+that implementations use "S256" when the device can handle sha256.
+Server must always support it anyway.
+
+Code challenge is generated just like the specification recommends,
+constructing 32-octet sequence that is then base64url encoded. This uses
+/dev/urandom to get randomness for code verifier which works well on
+Linux but might behave differently on other platforms and is not
+portable to non Unix-like OSs.
+
+Signed-off-by: Tomi Leppänen <tomi.leppanen@jolla.com>
+---
+ src/oauth2data.h     | 10 ++++++++++
+ src/oauth2plugin.cpp | 44 ++++++++++++++++++++++++++++++++++++++++++++
+ 2 files changed, 54 insertions(+)
+
+diff --git a/src/oauth2data.h b/src/oauth2data.h
+index 31184974f2cd25be7df5483415496a13f9df2cf0..7b4623b21a6678f7d32b4d4b1e8915191a058c29 100644
+--- a/src/oauth2data.h
++++ b/src/oauth2data.h
+@@ -125,6 +125,16 @@ namespace OAuth2PluginNS {
+          */
+         SIGNON_SESSION_DECLARE_PROPERTY(QString, TokenQuery);
+ 
++        /*!
++         * set to true to skip PKCE (RFC7636)
++         */
++        SIGNON_SESSION_DECLARE_PROPERTY(bool, SkipPKCE);
++
++        /*!
++         * code verifier for PKCE (RFC7636)
++         */
++        SIGNON_SESSION_DECLARE_PROPERTY(QByteArray, CodeVerifier);
++
+         /*!
+          * Any extra (non standard) parameters to send to the endpoint URI
+          */
+diff --git a/src/oauth2plugin.cpp b/src/oauth2plugin.cpp
+index a5f8e9769939ad1e0fe04840ef9208365167369b..452d03a88e927a26a5860a572efe98ef072486ec 100644
+--- a/src/oauth2plugin.cpp
++++ b/src/oauth2plugin.cpp
+@@ -25,6 +25,8 @@
+ #include "oauth2plugin.h"
+ #include "oauth2tokendata.h"
+ 
++#include <QCryptographicHash>
++#include <QFile>
+ #include <QJsonDocument>
+ #include <QUrl>
+ #include <QUrlQuery>
+@@ -68,6 +70,9 @@ const QString CLIENT_ID = QString("client_id");
+ const QString CLIENT_SECRET = QString("client_secret");
+ const QString REFRESH_TOKEN = QString("refresh_token");
+ const QString AUTH_ERROR = QString("error");
++const QString CODE_CHALLENGE = QString("code_challenge");
++const QString CODE_CHALLENGE_METHOD = QString("code_challenge_method");
++const QString CODE_VERIFIER = QString("code_verifier");
+ 
+ const QByteArray CONTENT_TYPE = QByteArray("Content-Type");
+ const QByteArray CONTENT_APP_URLENCODED = QByteArray("application/x-www-form-urlencoded");
+@@ -75,6 +80,33 @@ const QByteArray CONTENT_APP_JSON = QByteArray("application/json");
+ const QByteArray CONTENT_TEXT_PLAIN = QByteArray("text/plain");
+ const QByteArray CONTENT_TEXT_HTML = QByteArray("text/html");
+ 
++namespace {
++
++QByteArray getRandomCodeVerifier()
++{
++    /* Random code verifier generation as PKCE (RFC7636) defines it */
++    QByteArray data;
++    QFile urandom("/dev/urandom");
++    if (urandom.open(QIODevice::ReadOnly))
++        data = urandom.read(32);
```

Comment:
I don't know if I like this version more or less than the previous one. The previous one was portable, this isn't. I don't know if we really care about that. The previous one blocked (probably, at least with current implementation) if entropy was exhausted. This one doesn't block, but doesn't really work either. Like you wrote earlier, blocking shouldn't really be an issue for us. All in all, I think I like the previous version more.

Reply:
Ok. My problem with the previous implementation is that documentation of [random_device](https://en.cppreference.com/w/cpp/numeric/random/random_device) doesn't give any guarantees for the quality of the outputs. In gcc [libstdc++ the source code shows depending on build time configuration, it can be even the same as mt19937](https://gcc.gnu.org/onlinedocs/gcc-8.4.0/libstdc++/api/a00452_source.html#l01560). [gcc documentation](https://gcc.gnu.org/onlinedocs/gcc-8.4.0/libstdc++/api/a06295.html) doesn't say much what it would use in practice, but I guess that can be interpreted to be /dev/urandom on Unix-like systems, but there doesn't seem to be any guarantee. However, if we trust that on Sailfish OS that does get its randomness from kernel, I'm okay with sticking to that. I just think this one is a better implementation.

While I was reading about /dev/urandom use, I got the impression that it being non-blocking is not an issue when the use is shortly-lived, like in case of PKCE. Here is [one article that talks about it quite comprehensively](https://www.2uo.de/myths-about-urandom/), it's mostly fine, just the first section is formatted confusingly. So, AFAICT [it's perfectly fine](https://unix.stackexchange.com/questions/324209/when-to-use-dev-random-vs-dev-urandom) to [read /dev/urandom](https://crypto.stackexchange.com/questions/85533/what-issues-are-there-while-using-linuxs-dev-urandom-for-generating-cryptograp) for [a few bytes of randomness](https://security.stackexchange.com/questions/186086/is-always-use-dev-urandom-still-good-advice-in-an-age-of-containers-and-isola). And when comes to portability, well, yes, it only matters once we try to upstream these changes, but right now when this is used only on Sailfish OS it should not be a problem. Also I'm not sure what kind of targets this plugin has, if it actually supports something else than Unix-like.

Still, I would prefer getrandom() on Linux where it can be used, but in case of Sailfish OS, that's not an option. At least not on all devices and we'd need another version anyway for devices with 3.10 kernel.

Reply:
Adding another comment here as I want to address a few more points:

> The previous one blocked (probably, at least with current implementation) if entropy was exhausted.

How do you know? [AFAICS it also used /dev/urandom](https://github.com/gcc-mirror/gcc/blob/releases/gcc-8.3.0/libstdc%2B%2B-v3/include/tr1/random.h#L1461).

> This one doesn't block, but doesn't really work either.

What do you mean "doesn't really work"? If entropy is low, this will still get a random string of characters, but it can be more predictable.

> Like you wrote earlier, blocking shouldn't really be an issue for us.

I don't think I said that. I did say we probably won't be low on entropy as this will not happen early on boot and we won't need that much entropy (32 bytes is four 64-bit words or eight 32-bit words).

Reply:
Argh. I had my /dev/randoms mixed up, so feel free to ignore anything I wrote earlier.

Reply:
So, yeah, given that /dev/urandom really is the non-blocking random device, the only differences are: The old version is portable, the new one isn't. We don't know for certain what the old version does given any future gcc version, while we can be quite certain what the new version does. So, it's really a decision between portability and knowing what the implementation really does. I can live with both versions.

Reply:
> So, yeah, given that /dev/urandom really is the non-blocking random device, the only differences are: The old version is portable, the new one isn't. We don't know for certain what the old version does given any future gcc version, while we can be quite certain what the new version does. So, it's really a decision between portability and knowing what the implementation really does. I can live with both versions.

It would make sense to run the tests for this on a recent linux distro to make sure it will not break I think.

Reply:
> > So, yeah, given that /dev/urandom really is the non-blocking random device, the only differences are: The old version is portable, the new one isn't. We don't know for certain what the old version does given any future gcc version, while we can be quite certain what the new version does. So, it's really a decision between portability and knowing what the implementation really does. I can live with both versions.
> 
> It would make sense to run the tests for this on a recent linux distro to make sure it will not break I think.

What tests?

Reply:
Unless there is opposition, I'll merge this as it is later today.

One remark still, upstream might not care about older Qt versions, so they might want to use [QRandomGenerator](https://doc.qt.io/qt-5/qrandomgenerator.html) instead of C++'s `<random>`. We don't have that class as it was introduced in Qt 5.10. It would have the advantage that [QRandomGenerator::system()](https://doc.qt.io/qt-5/qrandomgenerator.html) is said to be cryptographically safe. Thus, it's quite plausible that they would go with a third solution.